File: leetcode024.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None:
            return None
        res = h = ListNode(-1)
        h.next = head
        while head != None:
            l2 = head
            head = head.next
            if head == None:
                return res.next
            l1 = head
            head = head.next
            h.next = l1
            l1.next = l2
            l2.next = head
            h = l2
        return res.next

File: leetcodec++/leetcode024.cpp
//
// Created by 黄敏辉 on 2019-06-29.
//

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode prehead(-1);
        prehead.next = head;

        ListNode *tmp1 = &prehead, *tmp2;

        while (tmp1->next && tmp1->next->next){
            tmp2 = tmp1->next;
            tmp1->next = tmp1->next->next;
            tmp2->next = tmp1->next->next;
            tmp1->next->next = tmp2;
            tmp1 = tmp1->next->next;
        }

        return prehead.next;
    }
};